src/VectorBundle.jl

```diff
@@ -0,0 +1,372 @@
+end
+
+vector_space_dimension(M::VectorBundleFibers{<:TCoTSpaceType}) = manifold_dimension(M.M)
+
+function vector_space_dimension(M::VectorBundleFibers{<:TensorProductType})
+    dim = 1
+    for space in M.VS.spaces
+        dim *= vector_space_dimension(VectorBundleFibers(space, M.M))
+    end
+    return dim
+end
+
+function representation_size(M::VectorBundleFibers{<:TCoTSpaceType})
+    representation_size(M.M)
+end
+
+"""
+    zero_vector!(M::VectorBundleFibers, v, x)
+
+Save the zero vector from the vector space of type `VS` at point `x`
+from manifold `M` to `v`.
+"""
+function zero_vector!(M::VectorBundleFibers, v, x)
+    error("zero_vector! not implemented for manifold $(typeof(M)), vector space of type $(typeof(VS)), vector of type $(typeof(v)) and point of type $(typeof(x)).")
+end
+
+function zero_vector!(M::VectorBundleFibers{<:TangentSpaceType}, v, x)
+    zero_tangent_vector!(M.M, v, x)
+    return v
+end
+
+"""
+    zero_vector(M::VectorBundleFibers, x)
+
+Compute the zero vector from the vector space of type `VS` at point `x`
+from manifold `M.M`.
+"""
+function zero_vector(M::VectorBundleFibers, x)
+    v = similar_result(M, zero_vector, x)
+    zero_vector!(M, v, x)
+    return v
+end
+
+"""
+    project_vector!(M::VectorBundleFibers, v, x, w)
+
+Project vector `w` from the vector space of type `VS` at point `x`
+and save the result to `v`.
+"""
+function project_vector!(M::VectorBundleFibers, v, x, w)
+    error("project_vector! not implemented for vector space manifold $(typeof(M)), vector space of type $(typeof(VS)), output vector of type $(typeof(v)) and input vector at point $(typeof(x)) with type of w $(typeof(w)).")
+end
+
+function project_vector!(M::VectorBundleFibers{<:TangentSpaceType}, v, x, w)
+    project_tangent!(M.M, v, x, w)
+    return v
+end
+
+"""
+    VectorBundle(M::Manifold, type::VectorSpaceType)
+
+Vector bundle on manifold `M` of type `type`.
+"""
+struct VectorBundle{TVS<:VectorSpaceType, TM<:Manifold} <: Manifold
+    type::TVS
+    M::TM
+    VS::VectorBundleFibers{TVS, TM}
+end
+
+function VectorBundle(VS::TVS, M::TM) where{TVS<:VectorSpaceType, TM<:Manifold}
+    return VectorBundle{TVS, TM}(VS, M, VectorBundleFibers(VS, M))
+end
+
+function representation_size(M::VectorBundle)
+    len_manifold = prod(representation_size(M.M))
+    len_vs = prod(representation_size(M.VS))
+    return (len_manifold + len_vs,)
+end
+
+TangentBundleFibers{M} = VectorBundleFibers{TangentSpaceType,M}
+TangentBundleFibers(M::Manifold) = VectorBundleFibers(TangentSpace, M)
+
+CotangentBundleFibers{M} = VectorBundleFibers{CotangentSpaceType,M}
+CotangentBundleFibers(M::Manifold) = VectorBundleFibers(CotangentSpace, M)
+
+TangentBundle{M} = VectorBundle{TangentSpaceType,M}
+TangentBundle(M::Manifold) = VectorBundle(TangentSpace, M)
+
+CotangentBundle{M} = VectorBundle{CotangentSpaceType,M}
+CotangentBundle(M::Manifold) = VectorBundle(CotangentSpace, M)
+
+"""
+    bundle_projection(M::VectorBundle, x::ProductRepr)
```

Comment:
Generally, I think `B` makes for a better default variable name  for bundles than `M`. It also changes `M.M` to `B.M`, which is a little nicer.